### DISCUSSION TITLE
"isNot" filter in "where" filter

### DISCUSSION THREAD

**Discussion by corymharper** [prisma/prisma#4340] [Q&A] [answered]
I've been looking all over the documentation, but I can't find it documented anywhere. Semantically it seems pretty straight forward:

```js
    const foos = await prisma.foo.findMany({
        where: {
            bar: {
                isNot: null,
            }
        },
    });
```

That this will return `foo` where related `bar` are not `null`, but I just wanted to confirm, is this the case?

**Comment by GeoffreyDick** [comment]
You can use `not` instead of `isNot`. It's under [Filtering](https://www.prisma.io/docs/concepts/components/prisma-client/filtering) in the documentation.

```javascript
    const foos = await prisma.foo.findMany({
      where: {
        bar: {
          not: null,
        },
      },
    });
```

**Reply by corymharper** [reply]
> You can use `not` instead of `isNot`. It's under [Filtering](https://www.prisma.io/docs/concepts/components/prisma-client/filtering) in the documentation.
> 
> ```js
>     const foos = await prisma.foo.findMany({
>       where: {
>         bar: {
>           not: null,
>         },
>       },
>     });
> ```

Hey there, thanks for the reply, interestingly, normally I would use `not`, but it's not showing up as a valid filter for the relationship in VSCode/According to Typescript. The two I am getting that seem like they might serve my purposes here are `isNot` and `NOT`. Maybe the type is in error?

**Reply by corymharper** [reply]
I'm generalizing the implementation I am working with here, but here are the generated types for prisma in this case:

`BarRelationFilter | BarWhereInput | null | undefined`

the relation filter looks like this:

```ts
type BarRelationFilter = {
  is?: BarWhereInput | null
  isNot?: BarWhereInput | null
}
```

and the where filter is shaped similarly to the following:

```ts
type BarWhereInput = {
  AND?: BarWhereInput | Enumerable<BarWhereInput>
  OR?: BarWhereInput | Enumerable<BarWhereInput>
  NOT?: BarWhereInput | Enumerable<BarWhereInput>
  id?: IntFilter | number
}
```

**Reply by GeoffreyDick** [reply]
Is `bar` a relation?

**Reply by corymharper** [reply]
> Is `bar` a relation?

Yes. There was also a typo throughout my example, I've fixed it

**Reply by GeoffreyDick** [reply]
Oh ok so you're basically querying every `foo` with a `bar`? I found Issue #2772 which might help. However I was unable to check against `null` values without getting an error. And using `undefined` gave me some strange results. This is what I tried:

```javascript
return prisma.foo.findMany({
  where: {
    bar: {
      // Try replacing "none" with "some" or "every" and see what happens. I got unexpected results
      none: {
        fooId: undefined,
      },
    },
  },
});
```

**Reply by ryands17** [reply, marked as answer]
@corymharper @GeoffreyDick 👋 
Assuming your schema is the following:

```prisma
model Foo {
  id    Int     @id @default(autoincrement())
  field String?
  bar   Bar[]
}

model Bar {
  id    Int     @id @default(autoincrement())
  field String?
  foo   Foo?    @relation(fields: [fooId], references: [id])
  fooId Int?
}
```

You can query for `bar` which are not null (having a relation) as follows:

```ts
await prisma.foo.findMany({
        where: { bar: { some: { fooId: { not: null } } } },
})
```

Let me know if this works.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
